crates/nu-command/src/misc/source.rs

```diff
@@ -41,15 +42,51 @@ impl Command for Source {
         call: &Call,
         input: PipelineData,
     ) -> Result<PipelineData, ShellError> {
-        // Note: this hidden positional is the block_id that corresponded to the 0th position
-        // it is put here by the parser
+        // Note: two hidden positionals are used here that are injected by the parser:
+        // 1. The block_id that corresponded to the 0th position
+        // 2. The block_id_name that corresponded to the file name at the 0th position
         let block_id: i64 = call.req_parser_info(engine_state, stack, "block_id")?;
+        let block_id_name: String = call.req_parser_info(engine_state, stack, "block_id_name")?;
         let block_id = BlockId::new(block_id as usize);
         let block = engine_state.get_block(block_id).clone();
+        let cwd = engine_state.cwd_as_string(Some(stack))?;
+        let pb = std::path::PathBuf::from(block_id_name);
+        let parent = pb.parent().unwrap_or(std::path::Path::new(""));
+        let file_path =
+            canonicalize_with(pb.as_path(), cwd).map_err(|err| ShellError::FileNotFoundCustom {
+                msg: format!("Could not access file '{}': {err}", pb.as_path().display()),
+                span: Span::unknown(),
+            })?;
+
+        let process_path = match pb.file_name() {
+            Some(name) => name.to_string_lossy().to_string(),
+            None => "unknown".to_string(),
+        };
+
+        // Add env vars so they are available to the script
+        stack.add_env_var(
+            "FILE_PWD".to_string(),
+            Value::string(parent.to_string_lossy(), Span::unknown()),
+        );
+        stack.add_env_var(
+            "CURRENT_FILE".to_string(),
+            Value::string(file_path.to_string_lossy(), Span::unknown()),
+        );
+        stack.add_env_var(
+            "PROCESS_PATH".to_string(),
+            Value::string(process_path, Span::unknown()),
+        );
 
         let eval_block_with_early_return = get_eval_block_with_early_return(engine_state);
 
-        eval_block_with_early_return(engine_state, stack, &block, input)
+        let val = eval_block_with_early_return(engine_state, stack, &block, input);
+
+        // After the script has ran, remove the env vars
```

Comment:
It should really reset them to their previous value, in case they were already set

Reply:
Good point. That should be fixed now. Thanks